File: backend/ride-go/internal/application/usecase/start_ride.go
package usecase

import (
	"errors"

	"github.com/leonardograselalmeida/fake_uber/internal/application/repository"
)

type StartRide struct {
	RideRepository repository.RideRepositoryInterface
}

type StartRideInput struct {
	RideId string
}

func (s *StartRide) Execute(input StartRideInput) error {
	ride, rideError := s.RideRepository.GetRideById(input.RideId)
	if rideError != nil {
		return rideError
	}
	if ride == nil {
		return errors.New("ride not found")
	}
	ride.Start()
	s.RideRepository.UpdateRide(ride)
	return nil
}

File: backend/ride-express/src/application/usecase/StartRide.ts
import type RideRepository from '../repository/RideRepository';

export default class StartRide {
  constructor(private rideRepository: RideRepository) {}

  async execute(input: Input) {
    const ride = await this.rideRepository.getById(input.rideId);
    if (!ride) throw new Error('Ride not found');
    ride.start();
    await this.rideRepository.update(ride);
  }
}

type Input = {
  rideId: string;
};
